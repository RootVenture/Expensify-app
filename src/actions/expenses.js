// import uuid from 'uuid';
import database from '../firebase/firebase';

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense,
});

// allowed because of redux thunk -- Create our data
export const startAddExpense = (expenseData = {}) => dispatch => {
  const { description = '', note = '', amount = 0, created = 0 } = expenseData;

  const expense = {
    description,
    note,
    amount,
    created,
  };
  return database
    .ref('expenses')
    .push(expense)
    .then(ref => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        })
      );
    });
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const startRemoveExpense = ({ id } = {}) => dispatch =>
  database
    .ref(`expenses/${id}`)
    .remove()
    .then(() => {
      dispatch(removeExpense({ id }));
    });

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});

export const startEditExpense = (id, updates) => dispatch =>
  database
    .ref(`/expenses/${id}`)
    .update(updates)
    .then(() => {
      dispatch(editExpense(id, updates));
    });

// SET_EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => dispatch =>
  // 1. fetch all expense data
  database
    .ref('expenses')
    .once('value')
    .then(snapshot => {
      // 2. parse all data into array
      const expeneses = [];

      snapshot.forEach(item => {
        expeneses.push({
          id: item.key,
          ...item.val(),
        });
      });
      // 3.dispatch SET_EXPENSES
      dispatch(setExpenses(expeneses));
    });
