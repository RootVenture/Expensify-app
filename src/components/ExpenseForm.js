import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));
class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toFixed(2).toString() : '',
      created: props.expense ? moment(props.expense.created) : moment(),
      calendarFocused: false,
      error: '',
    };
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description,
    }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    // regex to ensure user input is correct format
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount,
      }));
    }
  };

  onDateChange = created => {
    if (created) {
      this.setState(() => ({
        created,
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused,
    }));
  };

  onFormSubmit = e => {
    e.preventDefault(); // prevent page refresh

    if (!this.state.description || !this.state.amount) {
      // set error state
      this.setState(() => ({
        error: 'Please provide description and amount',
      }));
    } else {
      // clear the error
      this.setState(() => ({
        error: '',
      }));
      this.props.onSubmit({
        description: this.state.description,
        // convert to cents
        amount: parseFloat(this.state.amount, 10) * 100,
        // take our moment obj and convert to ms
        created: this.state.created.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        ExpenseForm
        <form onSubmit={this.onFormSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.created}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            // good for blocking days if booked
            isOutsideRange={day => false}
          />
          <textarea
            name="note"
            id="note"
            cols="30"
            rows="10"
            placeholder="Add a note for your expense (optional)"
            onChange={this.onNoteChange}
            value={this.state.note}
          />
          <button>Submit</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
