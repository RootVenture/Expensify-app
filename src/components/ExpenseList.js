import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getExpenses from '../selectors/expenses';

// export unconnected for test cases
export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div>
          <span className="list-item list-item--message">No expenses yet!</span>
        </div>
      ) : (
        props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
      )}
    </div>
  </div>
);

// utilize our selector for mapstatetoprops
const mapStateToProps = state => ({
  expenses: getExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseList);
