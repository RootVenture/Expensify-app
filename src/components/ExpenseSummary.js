import React, { Component } from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import { Link } from 'react-router-dom';
import selectExpensesTotal from '../selectors/expenses-total';
import getExpenses from '../selectors/expenses';

export class ExpenseSummary extends Component {
  render() {
    return (
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">
            Viewing <span>{this.props.expenseCount}</span>
            {this.props.expenseCount === 1 ? ' expense ' : ' expenses '}
            totaling <span>{numeral(this.props.expensesTotal / 100).format('$0,0.00')}</span>
          </h1>
          <div className="page-header__actions">
            <Link className="button" to="/create">
              Add Expense
            </Link>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const visibleExpenses = getExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
