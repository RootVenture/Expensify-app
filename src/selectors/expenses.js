import moment from 'moment';
// Filter for expenses
const getExpenses = (expenses, { text, sort, startDate, endDate }) =>
  expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.created);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(expense.created) : true;

      // figure out if expenses.description as the text variable string inside of it
      const textMatch =
        typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sort === 'date') {
        return a.created < b.created ? 1 : -1;
      } else if ((sort = 'amount')) {
        // expensive first
        return a.amount < b.amount ? 1 : -1;
      }
    });

export default getExpenses;
