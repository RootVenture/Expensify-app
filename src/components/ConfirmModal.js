import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startRemoveExpense } from '../actions/expenses';
import { history } from '../routers/AppRouter';

const ConfirmModal = props => {
  const removeExpense = () => {
    props.startRemoveExpense({ id: props.expense.id });

    // redirect to dashboard
    history.push('/');
  };
  return (
    <Modal
      isOpen={props.isOpen}
      onRequestClose={props.handleModalClose}
      contentLabel="Selected Option"
      closeTimeoutMS={200}
      className="modal"
    >
      <div className="modal__content">
        <h3>Are you sure you want to delete?</h3>
        <div className="modal__options">
          <button className="button" onClick={removeExpense}>
            Confirm
          </button>
          <button className="button button--secondary" onClick={props.handleModalClose}>
            Cancel
          </button>
        </div>
      </div>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startRemoveExpense: id => dispatch(startRemoveExpense(id)),
});
export default connect(undefined, mapDispatchToProps)(ConfirmModal);
