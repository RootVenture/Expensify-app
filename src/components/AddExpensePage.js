import React, { Component } from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    // redirect to our dashboard after submit
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
          <div className="content-container">
            <ExpenseForm onSubmit={this.onSubmit} />
          </div>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense)),
});

// need to connect store to access dispatch
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
