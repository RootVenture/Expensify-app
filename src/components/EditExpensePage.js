import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends Component {
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);

    // redirect to dashboard
    this.props.history.push('/');
  };

  onClick = id => {
    this.props.startRemoveExpense({ id: this.props.expense.id });

    // redirect to dashboard
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-title">{this.props.expense.description}</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button onClick={this.onClick} className="button button--secondary">
            Delete
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
