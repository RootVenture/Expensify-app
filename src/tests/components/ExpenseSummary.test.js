import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should render correctly with 1 expense', () => {
  const wrapper = shallow(<ExpenseSummary />);
  wrapper.setProps({
    expenseCount: 1,
    expensesTotal: 234,
  });

  expect(wrapper).toMatchSnapshot();
});

test('should render correctly with multiple expenses', () => {
  const wrapper = shallow(<ExpenseSummary />);
  wrapper.setProps({
    eexpenseCount: 2,
    expensesTotal: 345678,
  });

  expect(wrapper).toMatchSnapshot();
});
