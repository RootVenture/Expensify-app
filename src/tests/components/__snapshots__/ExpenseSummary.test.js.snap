// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       expense 
      totaling 
      <span>
        $2.34
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render correctly with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span />
       expenses 
      totaling 
      <span>
        $3,456.78
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
